crates/bevy_ecs/src/query/iter.rs

```diff
@@ -477,3 +479,61 @@ where
         }
     }
 }
+
+pub struct WithQuery<'world, 'state, I, Q, F>
+where
+    Q: WorldQuery,
+    F: WorldQuery
+{
+    iter: I,
+    query: &'world Query<'world, 'state, Q, F>,
+}
+
+impl<'world, 'state, I, Q, F> Iterator for WithQuery<'world, 'state, I, Q, F>
+where
+    I: Iterator<Item = Entity>,
+    Q: WorldQuery,
+    F: WorldQuery,
+{
+    type Item = <<Q as WorldQueryGats<'world>>::ReadOnlyFetch as Fetch<'world>>::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // if the entity iterator is done, this iterator is done
+        let mut entity = self.iter.next()?;
+        loop {
+            match self.query.get(entity).ok() {
+                Some(item) => return Some(item),
+                None => {
+                    entity = self.iter.next()?;
+                }
+            }
+        }
+    }
+}
+
+pub trait WithQueryExt {
```

Comment:
I don't like this naming convention. It's not really an extension trait: it's just a trait.

Not immediately sure on what to call it instead though.

Reply:
I renamed it `Joinable`, not strongly attached to it though, so any suggestions welcome